src/mlir/generator.cc

```diff
@@ -313,13 +298,9 @@ namespace mlir::verona
       auto name = AST::getID(nodes[0]);
       if (AST::isClassType(ast))
       {
-        // Pre-declaration multiple uses
-        if (auto type = typeTable.lookup(name))
-          return type;
-        // Class pre-declaration, cache it for multiple uses
-        auto type = ClassType::get(context, name);
-        typeTable.insert(name, type);
-        return type;
+        // Class pre-declaration, use `update` it for multiple uses
+        // before complete definition (done in `parseClass`).
+        return typeTable.update(name, ClassType::get(context, name));
```

Comment:
This is changing the behaviour a little as it keeps calling `ClassType::get` previously it would only be called once.

Would adding a `GetOrAdd` be more precise here, e.g. like in .NET 
[ConcurrentDictionary.getOrAdd](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netcore-3.1#System_Collections_Concurrent_ConcurrentDictionary_2_GetOrAdd__0_System_Func__0__1__).

Reply:
`ClassType::get` is a cheap constructor, similar to smart pointers, so should not be a problem calling it multiple types. It's also O(1). Previously, the additional search for an existing element would be at least O(logn).

It's not just "get or add" either, it's actually "insert or update with get at the end". I fear "GetOrAdd" would be misleading on updates.